File: Introduction_to_MPI/Exercises/fortran/ex3-ring/ring.f90
program ring

  use mpi
  implicit none

  integer :: ifail
  integer :: rank, size
  integer :: sum



  call mpi_init(ifail)

  call mpi_comm_rank(MPI_COMM_WORLD, rank, ifail)
  call mpi_comm_size(MPI_COMM_WORLD, size, ifail)

  ! Work out neighbouring ranks
  ! ...

  ! Initialise the sum
  sum = 0

  ! Send messages
  ! ...

  ! Display the result on all ranks, along with the correct answer

  write (unit = *, fmt = '(a,i3,a,i4,a,i4)') &
       "Rank ", rank, " has sum of ranks ", sum, " Answer ", (size - 1)*size/2

  call mpi_finalize(ifail)


end program ring

File: Introduction_to_MPI/Exercises/c/ex3-ring/ring.c
#include <mpi.h>
#include <stdio.h>

/* Ring program */

int main (int argc, char* argv[]){

  int rank, size;
  int sum = 0;

  /* ... */

  MPI_Init(&argc,&argv);

  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &size);

  /* Work out neighbouring ranks */
  /*... */

  /* Initialise the sum */
  /* ... */

  /* Send messages */
  /* ... */

  /* Display the result on all ranks, along with the correct answer */

   printf("Rank %2d has sum of ranks %d; Answer = %d\n",rank, sum, (size-1)*size/2);
  
  MPI_Finalize();

  return 0;

} 
